Add new platform EN75xx and device SmartFiber XP8421-B

cjdelisle opened the pull request:
EcoNet EN75xx is a big endian MIPS platform used in XPON (fiber), DSL, and SIM (3g/4g) applications. Complete GPL vender SDKs exist for this platform, but are based on Linux 2.6.

The SmartFiber XP8421-B is a fiber modem which is available for $20 online and has 512MB of memory, 256MB of SPI NAND flash and 2 USB 2.0 ports in addition to ethernet, wifi and XPON.

Many of the drivers for this SoC bear little resemblance to anything currently supported in OpenWRT or Linux upstream, so there is no realistic timeline for reaching a feature complete port. The aim of this patch is to provide the foundation (INTC, timer, SPI, NAND, image format) to be able to flash and boot to a shell.

I'm not getting paid for this, I'm just a hobbyist, and this is my first contribution. So any advice and guidance to help get this landed will be most appreciated.

The XP8421-B, and apparently many other devices of this platform, use a dual-image layout. I have chosen to reuse this to support dual-boot between OpenWRT and the factory firmware. Certain design decisions were made with the goal of not overwriting data that is used by the factory OS.

Flashing instructions (from bootloader):
1. Build and then locate the squashfs-tclinux.trx image file
2. Get the length of that file in hex: `printf '%X\n' "$(stat -c%s the-file-squashfs-tclinux.trx)"`
3. Connect to device with xmodem capability, e.g. `picocom --send-cmd lsx -vv -b 115200 /dev/ttyUSB0`
4. Switch device on and press a key within 3 seconds
5. Enter bootloader username and password: telecomadmin nE7jA%5m
6. Type: `xmdm 80020000 <file length hex>`
7. Quickly start xmodem and send the file, in picocom that is `ctrl+a ctrl+s <paste-the-file-name> enter` If the transfer fails to start, wait 30 seconds to a minute for the bootloader prompt to return and then try the command again.
8. Once the transfer has completed successfully, type the following `flash 80000 80020000 <file length hex>`
9. Type `go` or simply restart the device to boot into OpenWRT

cjdelisle committed:
kernel: Add new platform EN75xx and device SmartFiber XP8421-B

EcoNet EN75xx is a big endian MIPS platform used in XPON (fiber),
DSL, and SIM (3g/4g) applications. Complete GPL vender SDKs exist
for this platform, but are based on Linux 2.6.

The SmartFiber XP8421-B is a fiber modem which is available for
$20 online and has 512MB of memory, 256MB of SPI NAND flash and
2 USB 2.0 ports in addition to ethernet, wifi and XPON.

Many of the drivers for this SoC bear little resemblence to
anything currently supported in OpenWRT or Linux upstream, so
there is no realistic timeline for reaching a feature complete port.
The aim of this patch is to provide the foundation (INTC, timer,
SPI, NAND, image format) to be able to flash and boot to a shell.

I'm not getting paid for this, I'm just a hobbyist, and this is
my first contribution. So any advice and guidance to help get this
landed will be most appreciated.

The XP8421-B, and apparently many other devices of this platform,
use a dual-image layout. I have chosen to reuse this to support
dual-boot between OpenWRT and the factory firmware. Certain design
decisions were made with the goal of not overwriting data that is
used by the factory OS.

Flashing instructions (from bootloader):
1. Build and then locate the squashfs-tclinux.trx image file
2. Get the length of that file in hex:
	`printf '%X\n' "$(stat -c%s the-file-squashfs-tclinux.trx)"`
3. Connect to device with xmodem capability, e.g.
	`picocom --send-cmd lsx -vv -b 115200 /dev/ttyUSB0`
4. Switch device on and press a key within 3 seconds
5. Enter bootloader username and password: telecomadmin nE7jA%5m
6. Type: `xmdm 80020000 <file length hex>`
7. Quickly start xmodem and send the file, in picocom that is
	`ctrl+a ctrl+s <paste-the-file-name> enter`
    If the transfer fails to start, wait 30 seconds to a minute for
	the bootloader prompt to return and then try the command again.
8. Once the transfer has completed successfully, type the following
    `flash 80000 80020000 <file length hex>`
9. Type `go` or simply restart the device to boot into OpenWRT

Signed-off-by: Caleb James DeLisle <cjd@cjdns.fr>

cjdelisle commented:
@Ansuel This may be relevant to you and the Airoha target. In particular the [BMT/BBT](https://github.com/cjdelisle/openwrt/blob/new-platform-en75/target/linux/en75/files/drivers/mtd/nand/en75_bmt.c). I started out trying the BMT code from Airoha. My device does have some factory bad blocks, and it wiped the BBT which made a bit of a mess out of the disk. This implementation is quite conservative, by default it will not modify a BBT under any circumstances, only BMT, and it won't remap block in any partition unless it has been told to.

Djfe commented:
Noob here, but isn't it better to compare with the ramips implementation maybe?
https://github.com/openwrt/openwrt/blob/cbd1acbad3448280a180640e2a1a37a7144e1fb3/target/linux/ramips/dts/mt7621_linksys_ea7xxx.dtsi#L83 

I wonder how close these Econet MIPS chips compare to the older Trendchip-based designs like the RT6856F (how much they share).

cjdelisle commented:
@Djfe I thought the same thing when I started, but I was shocked by how little they have in common. Even simple things like clocksource and interrupt controller are entirely different from ramips. The BBT/BMT is also unfortunately different, I followed @Ansuel 's lead and hung it off of the mtk_bbt framework but it doesn't match any of the other mtk_bbt algorithms. GPL dump archeology indicates that this is indeed a descendent of Trendchip, I see a lot of TC3162 mentioned.

On another note, it seems I baffed the commit message, is it appropriate to just do a push-f ?

Djfe commented:
RT6856F isn't ramips, that's why it was never supported by Linux. It's the old trendchip architecture. 

It was also sometimes named TC3262 OR MT6856F, I think.
see https://wikidevi.wi-cat.ru/TrendChip

I'm not sure if this is even helpful. but I thought maybe there is old GPL sourcecode for routers like the d-link DIR-845L. maybe that source code contains helpful comments and/or further relevant info on the platform (if it has similarities)

It's actually recommended to do git rebase and force pushes on Pull Requests. Feel free to do so :)
You may also mark the PR as draft status until you feel like it is ready to be committed into the main branch.

cjdelisle commented:
The reason I named this en75 rather than tc3162 is because I wanted people to with en75 devices to know that this is for them, and tc3162 is so old by now that it's not that interesting anymore, and even if it was, I wouldn't know where to find one.

I fixed the commit message header, and as far as I know this should be good to merge so right now I'm looking for reviews.

Notupus commented:
Hey, I dont have this specific device but I have a Zyxel VMG8623-T50B, its a real common EN7516 iad and zyxel provides GPL sources for it if it can be useful for this. If this gets merged I will likely soon make a pr for it since it should be pretty much the same to support the basics.

cjdelisle commented:
@Notupus Right now I intend to focus on full support for the EN751221 family, but adding the EN751627, EN7528, and EN8580 should not be that difficult.

alexceltare2 commented:
Any core dev willing to review and merge this so we can get the ball rolling?

svanheule commented:
setup.c looks pretty minimal, so this can probably be reduced to a target derived from MIPS_GENERIC. Submit the code upstream first for a thorough review.

We already have airoha/en7523 (an ARM target), so naming this one en75 might become confusing too.

cjdelisle commented:
Hello @svanheule , thank you for looking at this.

The main issues with making this a board in MIPS_GENERIC are:
1. The bad old bootloader I'm dealing with requires that I compress the kernel with zboot, and that I load the zboot bin at 80020000 so the load address needs to be moved well above MIPS_GENERIC's 1MB offset
2. MIPS_GENERIC_KERNEL selects CPU_MIPSR2_IRQ_EI and CPU_MIPSR2_IRQ_VI, and when EI/VI is enabled, this INTC only sends vectored interrupts which causes `mti,cpu-interrupt-controller` to misbehave[1].
3. MIPS_GENERIC_KERNEL selects CEVT_R4K which probably exists, but I never tried much to get it working because all existing vendor code uses a 200mhz "High Performance Timer" instead.

Of these issues, 3 is probably solvable, 1 requires adding a custom load address, but 2 seems to be a bit of a deal-breaker unless there's a trick to stop the processor sending vectored interrupts only.

Regarding the name "en75" I'm open to other suggestions, I like that it's short enough to prefix a symbol but if there's something else better, I don't mind changing it. This is the same issue as mediatek / mtmips, sequential chip numbers with different architectures practically guarantee confusion. It *seems* that Airoha is leaning toward exclusively using ARM, and toward rebranding with the "an75" prefix, so there is some hope that we might reach a point where an75 = ARM and en75 = MIPS[2].

Can you advise me what is the best next step to move this toward being able to merge to OpenWRT ?

---

[1]: I wrote a handler for vectored interrupts but decided not to contribute it because it's bad code. I may in the future consider contributing a generic vector handler which separates handling from the controller, but that's a different topic.
[2]: I don't know anyone at Airoha so I don't know if this is their plan, I'm just basing this off of what I have observed.

abajk commented:
> Of these issues, 3 is probably solvable, 1 requires adding a custom load address, but 2 seems to be a bit of a deal-breaker unless there's a trick to stop the processor sending vectored interrupts only.

Sending code upstram is a very good thing. You can get a greater feedback from people familiar with MIPS.

svanheule commented:
> 1. The bad old bootloader I'm dealing with requires that I compress the kernel with zboot, and that I load the zboot bin at 80020000 so the load address needs to be moved well above MIPS_GENERIC's 1MB offset

I see you're already using `ZBOOT_LOAD_ADDR` for this. Note that I'm not a MIPS expert, just basing this on my experience with the Realtek upstreaming. Upstream would be able to give you better feedback.

> 2\. MIPS_GENERIC_KERNEL selects CPU_MIPSR2_IRQ_EI and CPU_MIPSR2_IRQ_VI, and when EI/VI is enabled, this INTC only sends vectored interrupts which causes `mti,cpu-interrupt-controller` to misbehave[1].

If this is truly broken, upstream will allow you to not use MIPS_GENERIC, I suppose. Alternatively, they will be able to provide you with the correct pointers on how to properly handle things.

> 3\. MIPS_GENERIC_KERNEL selects CEVT_R4K which probably exists, but I never tried much to get it working because all existing vendor code uses a 200mhz "High Performance Timer" instead.

AFAIK CEVT_R4K runs at (half?) the CPU clock, so that should be high performance too. Should come pretty much for free with MIPS_GENERIC.

Naming can then be decided on after the platform code has been reviewed. Mediatek's ARM chips go under the "mediatek" target, while their MIPS chips go under the "ramips" target (from their acquisition of Ralink). Same could be done here: "airoha" for the current ARM chips, and "econet"/"enmips"/... for the older econet designs.

The quickest way to get a new platform accepted in OpenWrt is to have it merged upstream. The realtek target, for example, carries a lot of downstream code and every kernel upgrade is a pain.

cjdelisle commented:
Thank you for the advice.

> I see you're already using ZBOOT_LOAD_ADDR for this.

ZBOOT_LOAD_ADDR is 80020000 per bootloader requirement. The actual load address can be whatever we want but it has to be above 80020000 + the maximum possible size of compressed kernel + 4MB ZBOOT heap, so 80100000 is a no-go. That said, if the maintainers of MIPS_GENERIC don't mind a special-case load address, that's good enough for me.

> AFAIK CEVT_R4K runs at (half?) the CPU clock, so that should be high performance too. Should come pretty much for free with MIPS_GENERIC.

You raise a good point. All I can say is that every vendor implementation (there are a few) has gone out of their way to use this clock, to the point of even patching cevt-r4k.c to add hooks for it. These chips are common in fiber modems which require tight timing for TDMA so I can only conjecture that the CPU counter doesn't have what is needed.

I can go back and try get CEVT_R4K working. If it works then I at least don't need to submit the HPT in my first patchset. I'm thinking I should include the intc and (hopefully) CEVT / CSRC_R4K so that upstream alone can boot to a console with initramfs. I'll exclude the SPI driver (which I lifted from /airoha) and the NAND BMT because the BMT uses a MediaTek BMT framework[1] which is not upstreamed. I'll make the first patchset create a new platform and explain the issues with using MIPS_GENERIC, if someone sees it and says "oh you can just do this" then I can re-submit as a board under generic. Does that sound like the right approach?  (Thanks for the hand-holding, I'm new here)

[1]: It would be nice to upstream a generic MTD->MTD translator which provides a clean framework for any BMT driver to hook into, but I don't want to block EcoNet support on it.

cjdelisle committed:
en75: Add support for TP-Link Archer VR1200v (v2) and Generic EN751221

Add support for TP-Link Archer VR1200v (v2), a DSL modem based
on the EN7513G SoC. Installation instructions can be found here:
https://openwrt.org/inbox/toh/tp-link/archer_vr1200v

Also added a "Generic EN751221" initramfs image for testing and
recovery of non-specific EN751221 devices.

Signed-off-by: Caleb James DeLisle <cjd@cjdns.fr>

cjdelisle committed:
en75: spi: Quirk to fix lost writes on SPI NAND

The EN751221 SPI controller requires quiet time between operations
otherwise NAND writes are often (almost always) not committed.
udelay(100) works but does not account for the SPI clock speed which
is configurable. This issue has been observed on two different SoCs
with different brands of SPI NAND. The vendor code sets the CS state
twice in order to add one SPI op worth of quiet time before and
after each operation, this seems like a good strategy and is in any
case well tested.

Signed-off-by: Caleb James DeLisle <cjd@cjdns.fr>

cjdelisle committed:
en75: Add utility to change OS A/B boot flag in flash

EN75xx devices typically have dual OS for redundancy, we repurpose
this to provide "dual boot" between OpenWRT and Factory OS. This
patch adds a utility to change the flag in flash which tells the
bootloader which OS to boot to.

Signed-off-by: Caleb James DeLisle <cjd@cjdns.fr>

cjdelisle commented:
Update: After testing every CPU interrupt line, I can now say with some confidence that this SoC does *not* have CEVT_R4K.

I will start the process of getting a basic "initramfs boot to a shell" port submitted upstream, but in the mean time I have just pushed some improvements and a new device [TP-Link Archer VR1200v (v2)](https://openwrt.org/inbox/toh/tp-link/archer_vr1200v) to my PR.

cjdelisle commented:
Closing this for the moment because I'm sending a patchset upstream so pending merge of that, this PR will change dramatically.

alexceltare2 commented:
> Closing this for the moment because I'm sending a patchset upstream so pending merge of that, this PR will change dramatically.

Hope it gets approved 🫡

Djfe commented:
Good Luck :)

unixty commented:
> This was re-opened as #19021 and was merged last June.

Good job, so en7516gt keenetic kn-3610 can be added to openwrt? I want to try contribute in project. 
I see that a few devices just tested on this platform
